server/src/main/java/org/opensearch/common/cache/stats/CacheStats.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.opensearch.core.common.io.stream.Writeable;
+
+/**
+ * Interface for access to any cache stats. Allows accessing stats by dimension values.
+ * Stores an immutable snapshot of stats for a cache. The cache maintains its own live counters.
+ */
+public interface CacheStats extends Writeable {// TODO: also extends ToXContentFragment (in API PR)
+
+    // Method to get all 5 values at once
+    CacheStatsResponse.Snapshot getTotalStats();
+
+    // Methods to get total values.
+    long getTotalHits();
+
+    long getTotalMisses();
+
+    long getTotalEvictions();
+
+    long getTotalSizeInBytes();
+
+    long getTotalEntries();
+}
```

Comment:
Could this interface (and therefore `MultiDimensionCacheStats`) just be the tree of all stats by dimension?

The `CacheStats` API would look like:

```
long getHits();
long getMisses();
long getEvictions();
long getSizeInBytes();
long getEntryCount();

Map<String, CacheStats> getChildren();
```

Then `MultiDimensionCacheStats` would look like:

```

final Map<String, CacheStats> children; // Passed to constructor

public MultiDimensionCacheStats(Map<String, CacheStats> children) {
  this.children = Collections.unmodifiableMa(children);
}

public Map<String, CacheStats> getChildren() {
  return children
}

public long getHits() {
  long hits = 0;
  for (CacheStats child : children.values()) {
    hits += child.getHits();
  }
  return hits;
}

// Other methods

```

Then what's currently called `CacheStatsResponse.Snapshot` could look like:

```
public static class Snapshot implements CacheStats {

final long hits;
final long misses;
final long evictions;
final long sizeInBytes;
final long entryCount;

public Map<String, CacheStats> getChildren() {
  return Collections.emptyMap();
}

public long getHits() {
  return hits;
}

// ... other getters ...

```

If the `StatsHolder` were stored as a recursive tree, life would be a lot easier. If it's not, then I guess you'd need something like the following:

```

private static class KeyTree {
  Map<String, KeyTree> children = new HashMap<>();
}

public CacheStats snapshot() {
  List<Key> keys = new ArrayList<>(statsMap.keySet()); // Save a copy of the keys, so we can slice them.  
  KeyTree keyTree = new KeyTree();
  for (Key key : keys) {
    KeyTree current = keyTree;
    for (String dimensionValue :  key.dimensionValues) {
      keyTree.computeIfAbsent(dimensionValue, (k) -> KeyTree::new);
    }
  }
  return buildRecursiveCacheStats(currentNode, collections.emptyList());
}

private CacheStats buildRecursiveCacheStats(KeyTree currentNode, List<String> keyPath) {
  if (currentNode.children.isEmpty()) {
    // Leaf node
    return statsMap.get(new Key(keyPath)).snapshot();
  }
  // Inner node
  Map<String, CacheStats> stats = new TreeMap<>();
  for (Map.Entry<String, KeyTree> entry : currentNode.children.entrySet()) {
    List<String> newKeyPath = new ArrayList<>(keyPath);
    newKeyPath.add(entry.getKey());
    stats.put(entry.getKey(), buildRecursiveCacheStats(entry.getValue(), newKeyPath);
  }
  return stats;
}
```

The whole `levels` thing seems kind of unnecessary when you can just return the tree.

Note that the above only works if every element in the `StatsHolder` has the same dimensionality (though that's already enforced as an invariant -- right?)

Reply:
I agree that we could do this, but is the current logic unacceptable? I've already reimplemented the followup TieredSpilloverCacheStats PR and API PR logic based on the changes we've made to this PR in the last few days and I would have to entirely rewrite those. We would still have to re-process the tree based on levels in MultiDimensionCacheStats.ToXContent() anyway. I'm not totally sure I see the benefit in making this change.

Reply:
IMO, the current logic is hard to follow. 

The dimensions naturally organize the stats into a tree (broken down by the first dimension, then the second dimension, etc.). The JSON output in the stats response is going to be a tree. Why not just build a tree? The `toXContent` output would just be a recursive tree traversal.